internal/config/instance.go

```diff
@@ -113,65 +117,68 @@ func Get() (*Instance, error) {
 }
 
 // Set sets a value at the given key
-func (i *Instance) Set(key string, value interface{}) {
+func (i *Instance) Set(key string, value interface{}) error {
 	i.rwLock.Lock()
 	defer i.rwLock.Unlock()
-	i.viper.Set(key, value)
+
+	err := i.ReadInConfig()
+	if err != nil {
+		return err
+	}
+
+	i.data[strings.ToLower(key)] = value
+
+	err = i.save()
+	if err != nil {
+		return err
+	}
+
+	return nil
 }
 
-// GetString retrieves a string for a given key
-func (i *Instance) GetString(key string) string {
+func (i *Instance) get(key string) interface{} {
 	i.rwLock.RLock()
 	defer i.rwLock.RUnlock()
-	return i.viper.GetString(key)
+	return i.data[strings.ToLower(key)]
 }
 
+// GetString retrieves a string for a given key
+func (i *Instance) GetString(key string) string {
+	return cast.ToString(i.get(key))
+}
+
+// AllKeys returns all of the curent config keys
 func (i *Instance) AllKeys() []string {
-	i.rwLock.RLock()
-	defer i.rwLock.RUnlock()
-	return i.viper.AllKeys()
+	var keys []string
+	for k := range i.data {
+		keys = append(keys, k)
+	}
+	return keys
 }
 
 // GetStringMapStringSlice retrieves a map of string slices for a given key
 func (i *Instance) GetStringMapStringSlice(key string) map[string][]string {
-	i.rwLock.RLock()
-	defer i.rwLock.RUnlock()
-	return i.viper.GetStringMapStringSlice(key)
-}
-
-// SetDefault sets the default value for a given key
-func (i *Instance) SetDefault(key string, value interface{}) {
-	i.rwLock.Lock()
-	defer i.rwLock.Unlock()
-	i.viper.SetDefault(key, value)
+	return cast.ToStringMapStringSlice(i.get(key))
 }
 
 // GetBool retrieves a boolean value for a given key
 func (i *Instance) GetBool(key string) bool {
-	i.rwLock.RLock()
-	defer i.rwLock.RUnlock()
-	return i.viper.GetBool(key)
+	return cast.ToBool(i.get(key))
```

Comment:
This isn't on you, @MDrakos (I think), but not returning an error from these functions seems haphazard. Since this is being worked on now, I suggest that it should be added. If not now, then a story filed. Calling code can continue ignoring the error for a while, with another story (or multiple stories) to handle upgrading those call sites.

Reply:
I agree that ignoring the error here could cause issues. One of the ACs of this story is that the config API remains unchanged. I've already violated this AC on the `Set` calls but that was after discussion with Nathan and concluding that we are not using `Set` in many places. Updating the get functions would be a larger undertaking that I don't think should be in the scope of this story but something worth visiting in the future.